### PR TITLE
Calcite SQL compiler syntax support for order by expression and set LENIENT SQL conformance

### PR DESCRIPTION
**Pull request by fx19880617**
1. To support query with order by expressions like:
```
select group_city, sum(rsvp_count), count(*) from meetupRsvp group by group_city order by sum(rsvp_count), count(*) limit 10
```
2. Use LIMIT with GROUP BY to display records.

3. Set SQL conformance LENIENT to make below query syntax works:
- Support `LIMIT offset,count` in Calcite Parser
- Support `!=` as not equal in Calcite Parser

### CODE REVIEW

File: pinot-common/src/main/java/org/apache/pinot/sql/parsers/CalciteSqlParser.java [hunk 1]
```diff
@@ -224,17 +224,16 @@ private static Expression convertOrderBy(SqlNode node) {
...
-      case IDENTIFIER:
-        expression = RequestUtils.getFunctionExpression("ASC");
-        expression.getFunctionCall().addToOperands(toExpression(node));
-        break;
       case DESCENDING:
         SqlBasicCall basicCall = (SqlBasicCall) node;
         expression = RequestUtils.getFunctionExpression("DESC");
         expression.getFunctionCall().addToOperands(toExpression(basicCall.getOperands()[0]));
         break;
+      case IDENTIFIER:
```
Comment (kishoreg) [inline pinot-common/src/main/java/org/apache/pinot/sql/parsers/CalciteSqlParser.java]: Should we add ascending as well for readability?

Comment (fx19880617) [reply pinot-common/src/main/java/org/apache/pinot/sql/parsers/CalciteSqlParser.java]: there is no ascending key word in calcite sql

Comment (fx19880617) [reply pinot-common/src/main/java/org/apache/pinot/sql/parsers/CalciteSqlParser.java]: basically by default, it's ascending already

**Comment by Jackie-Jiang** [comment]
We don't support such queries yet, so what is the current behavior with this pr checked in? Will the engine throw exception or just ignore the order by? Throwing exception is better IMO.

**Comment by fx19880617** [comment]
> We don't support such queries yet, so what is the current behavior with this pr checked in? Will the engine throw exception or just ignore the order by? Throwing exception is better IMO.

@npawar added the support for group by order by aggregation functions.

File: pinot-common/src/main/java/org/apache/pinot/sql/parsers/CalciteSqlParser.java [hunk 2]
```diff
@@ -111,6 +112,7 @@ private static void setOptions(PinotQuery pinotQuery, List<String> optionsStatem
   private static PinotQuery compileCalciteSqlToPinotQuery(String sql) {
     SqlParser.ConfigBuilder parserBuilder = SqlParser.configBuilder();
     parserBuilder.setLex(PINOT_LEX);
+    parserBuilder.setConformance(SqlConformanceEnum.LENIENT);
```
Comment (siddharthteotia) [inline pinot-common/src/main/java/org/apache/pinot/sql/parsers/CalciteSqlParser.java]: Why is this needed?

Comment (fx19880617) [reply pinot-common/src/main/java/org/apache/pinot/sql/parsers/CalciteSqlParser.java]: I think I committed another change for supporting `LIMIT offset,numRecords` into this.

Comment (fx19880617) [reply pinot-common/src/main/java/org/apache/pinot/sql/parsers/CalciteSqlParser.java]: basically calcite parser cannot parse `LIMIT X,Y` until this conformance is added.

**Comment by siddharthteotia** [comment]
This is supported on all paths right?

(1) PQL -> PqlCompiler -> BrokerRequest
(2) PQL -> PqlCompiler -> BrokerRequest -> PinotQuery -> brokerRequest.setPinotQuery()
(3) SQL -> Calcite parser -> PinotQuery

File: pinot-common/src/main/java/org/apache/pinot/sql/parsers/CalciteSqlParser.java (hunk 2 above)
Comment (fx19880617) [reply pinot-common/src/main/java/org/apache/pinot/sql/parsers/CalciteSqlParser.java]: > This is supported on all paths right?
> 
> (1) PQL -> PqlCompiler -> BrokerRequest
> (2) PQL -> PqlCompiler -> BrokerRequest -> PinotQuery -> brokerRequest.setPinotQuery()
> (3) SQL -> Calcite parser -> PinotQuery

For path 3 it's SQL -> Calcite parser -> PinotQuery -> BrokerRequest
Query processing still depends on BrokerRequest.

Also plan to make query console default to go through path 3 in PR https://github.com/apache/incubator-pinot/pull/4994

**Comment by fx19880617** [comment]
> This is supported on all paths right?
> 
> (1) PQL -> PqlCompiler -> BrokerRequest
> (2) PQL -> PqlCompiler -> BrokerRequest -> PinotQuery -> brokerRequest.setPinotQuery()
> (3) SQL -> Calcite parser -> PinotQuery

For path 3 it's SQL -> Calcite parser -> PinotQuery -> BrokerRequest
Query processing still depends on BrokerRequest.

Also plan to make query console default to go through path 3 in PR #4994

File: pinot-common/src/main/java/org/apache/pinot/sql/parsers/CalciteSqlParser.java (hunk 2 above)
Comment (npawar) [reply pinot-common/src/main/java/org/apache/pinot/sql/parsers/CalciteSqlParser.java]: @siddharthteotia we did an exercise at LI wherein we passed all possible used queries though Calcite parser. We caught certain things which do not pass under the default sql conformance level (`!=` and `LIMIT offset, count`). One of the solutions for that we discussed was to set the conformance to be more lenient.

**Comment by npawar** [comment]
This LGTM.
I see you've also included the change to use LIMIT in group by queries. Please call that out as well in the description.

**Comment by siddharthteotia** [comment]
> > This is supported on all paths right?
> > (1) PQL -> PqlCompiler -> BrokerRequest
> > (2) PQL -> PqlCompiler -> BrokerRequest -> PinotQuery -> brokerRequest.setPinotQuery()
> > (3) SQL -> Calcite parser -> PinotQuery
> 
> For path 3 it's SQL -> Calcite parser -> PinotQuery -> BrokerRequest
> Query processing still depends on BrokerRequest.

Just to be on the same page (and this is not really related to this PR), for path 3, once we have the PinotQuery, we are only going to set it inside BrokerRequest -- we are not going to build a full fledged broker request all together by converting PinotQuery to BrokerRequest. 

So the desired end state here is to have the BrokerRequest as just a wrapper thrift definition over PinotQuery and the servers should build their execution plan from the information stored in brokerRequest.getPinotQuery() -- this should be sufficient to build the plan.

In fact, since PinotQuery is a thrift object on its own, we should be sending that to the servers and migrate each and every operator to work off PinotQuery -> QueryContext as described in this doc

**Comment by fx19880617** [comment]
> This LGTM.
> I see you've also included the change to use LIMIT in group by queries. Please call that out as well in the description.

sure, I think also for SQL parser, we actually don't support TOP, so LIMIT is used.

**Comment by fx19880617** [comment]
> > > This is supported on all paths right?
> > > (1) PQL -> PqlCompiler -> BrokerRequest
> > > (2) PQL -> PqlCompiler -> BrokerRequest -> PinotQuery -> brokerRequest.setPinotQuery()
> > > (3) SQL -> Calcite parser -> PinotQuery
> > 
> > 
> > For path 3 it's SQL -> Calcite parser -> PinotQuery -> BrokerRequest
> > Query processing still depends on BrokerRequest.
> 
> Just to be on the same page (and this is not really related to this PR), for path 3, once we have the PinotQuery, we are only going to set it inside BrokerRequest -- we are not going to build a full fledged broker request all together by converting PinotQuery to BrokerRequest.
> 
> So the desired end state here is to have the BrokerRequest as just a wrapper thrift definition over PinotQuery and the servers should build their execution plan from the information stored in brokerRequest.getPinotQuery() -- this should be sufficient to build the plan.
> 
> In fact, since PinotQuery is a thrift object on its own, we should be sending that to the servers and migrate each and every operator to work off PinotQuery -> QueryContext as described in this doc

Currently we are still convert PinotQuery to internal BrokerRequest fields objects. The end goal is to use fields inside PinotQuery for internal query processing.

File: pinot-common/src/main/java/org/apache/pinot/sql/parsers/CalciteSqlParser.java (hunk 2 above)
Comment (siddharthteotia) [reply pinot-common/src/main/java/org/apache/pinot/sql/parsers/CalciteSqlParser.java]: Got it. Thanks

File: pinot-common/src/main/java/org/apache/pinot/pql/parsers/pql2/ast/SelectAstNode.java [hunk 3]
```diff
@@ -31,12 +31,12 @@
  * AST node for a SELECT statement.
  */
 public class SelectAstNode extends BaseAstNode {
+  public static final int DEFAULT_RECORD_LIMIT = 10;
```
Comment (mayankshriv) [inline pinot-common/src/main/java/org/apache/pinot/pql/parsers/pql2/ast/SelectAstNode.java]: Is it possible to just have one variable for TOP/LIMIT? Or does LIMIT imply that there may not be any sorting if ORDERBY not provided?

Comment (siddharthteotia) [reply pinot-common/src/main/java/org/apache/pinot/pql/parsers/pql2/ast/SelectAstNode.java]: Yes LIMIT implies no sorting/ordering unless ORDER BY is specified. In fact in MySql, both LIMIT and TOP are synonymous. 

I think we should eventually move to LIMIT -- now that we have GROUP BY with ORDER BY and are also converging towards standard SQL, we should have two variants (1) LIMIT and (2) LIMIT with ORDER BY

Comment (mayankshriv) [reply pinot-common/src/main/java/org/apache/pinot/pql/parsers/pql2/ast/SelectAstNode.java]: Wondering what is a good use case for non ordered response? And/or if we should consider a default ordering?
